[GSoC] Added 240-EfficientDet-optimization-notebook

ashish-2005 opened the pull request:
#### Prerequisite of GSoc'23.
This notebook demonstrate usage of EfficientDet model with openVINO IR and post-training optimization with NNCF

ashish-2005 commented:
my notebook is passing windows and ubuntu tests but failing on macos-11,12 and i am unable to find why , @adrianboguszewski can you help me to find the cause , It would be a great help

adrianboguszewski commented:
Hi @ashish-2005, before we understand what's wrong with macOS, we have to fix other issues:
- you don't have to clone the entire automl repo; I recommend you use tensorflow_hub to get the EfficientDet model
- you don't have to install any requirement inside the notebook if it's already covered in the requirements.txt file
- please look at notebook_utils.py because some utilities are already implemented there (e.g. load_image, download_file functions)
- please use model optimizer python API, as it's been shown in 226
- please check your grammar, interpunction and typos in any automatic tool (e.g., a free account of Grammarly)
- the notebooks should cover all important topics and at the same time should be minimal

ashish-2005 commented:
Thanks for your feedback , I will resolve the mentioned issues as soon as possible and commit the updates

ashish-2005 commented:
Hii,
Apologies for the late update 
There were some issues in openVINO API for tensorflow SavedModel format which I discussed with Roman [here](https://github.com/openvinotoolkit/openvino/issues/16095)
Updates Made :
* replaced downloading the whole EfficientDet repository with TensorFlow-Hub API
* replaced command line mo tool with openVINO python API 

Thanks

*I would request not to run workflow for now as I am still making changes*

adrianboguszewski commented:
Please also do the self-check:
- [ ] each function is described by docstrings and type hints
- [ ] notebook contains explicit descriptions and explanatory diagrams
- [ ] there is a README.md file in consistent style (look at other notebooks)
- [ ] the notebook is added to the main README and notebooks/README files
- [ ] there are no grammar, interpunctions or typo issues (use any free tool for that e.g. Grammarly)
- [ ] there are no committed files besides notebook and readme (please use images or videos from data dir)
- [ ] all CI checks passed

ashish-2005 commented:
### Self-check Update
* [x] each function is described by docstrings and type hints
* [x] notebook contains explicit descriptions and explanatory diagrams
* [x] the notebook doesn't use any data (image, video, etc.) that is not CC4.0 licensed
* [x]  there is a README.md file in consistent style 
* [x] the notebook is added to the main README and notebooks/README files
* [x] there are no grammar, interpunctions or typo issues 
* [x] there are no committed files besides notebook and readme
* [ ] all CI checks passed

All CI have passed except `docker_nbval` due to [227 notebook](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/227-whisper-subtitles-generation) which I guess has been handled in PR #929

adrianboguszewski commented on notebooks/240-EfficientDet-object-detection/README.md:
```diff
@@ -0,0 +1,23 @@
+# EfficientDet Optimization with OpenVINO
+Research Paper : [EfficientDet: Scalable and Efficient Object Detection](https://openaccess.thecvf.com/content_CVPR_2020/papers/Tan_EfficientDet_Scalable_and_Efficient_Object_Detection_CVPR_2020_paper.pdf)
+<a href='https://github.com/google/automl/tree/master/efficientdet'>EfficientDets</a> are a family of object detection models, which achieve state-of-the-art 55.1mAP on COCO test-dev, yet being 4x - 9x smaller and using 13x - 42x fewer FLOPs than previous detectors. Our models also run 2x - 4x faster on GPU, and 5x - 11x faster on CPU than other detectors
```
HTML tags should be replaced by Markdown. The paragraph cannot state "our model" as this is not ours. It should be rewritten to describe EfiicientDet as an external model.

adrianboguszewski commented on notebooks/240-EfficientDet-object-detection/README.md:
```diff
@@ -0,0 +1,23 @@
+  * Validating Inference of INT8 model
+  * Evaluating performance of INT8 model by AP
+
+<img src='https://user-images.githubusercontent.com/71766106/226086430-a7e3cdc4-1f99-4c46-89f9-60dcbadea44a.png' >
```
The image (in Markdown, not HTML) must be placed at the top. See other readme files.

adrianboguszewski commented on notebooks/240-EfficientDet-object-detection/README.md:
```diff
@@ -0,0 +1,23 @@
+
+* Downloading and exporting the efficientDet-d0 from TensorFlow-HUB
+  * Validating inference of tf-model
+* Converting tf-model to OpenVINO-IR(Intermediate representation) file
+  * Validating inference of OpenVINO-model
+  * Evaluating performance by AP( average Precision)
```
Still interpunction

adrianboguszewski replied:
I mean spaces

adrianboguszewski reviewed (changes requested):
Is this PR finished? I see sections like "Will be included later". Does it mean you still work on this?

And please:
- rebase your work to include the #929 fix
- change the name of your notebook to follow the adopted style (lowercase only, no underscores)

ashish-2005 commented:
> Is this PR finished? I see sections like "Will be included later". Does it mean you still work on this?
> 
> And please:
> 
> * rebase your work to include the #929 fix
> * change the name of your notebook to follow the adopted style (lowercase only, no underscores)

Hii @adrianboguszewski 
The "Will be included later" part is model optimization with NNCF which I am not sure whether to add or not, as my notebook already takes a lot of time to run because of the big tf-hub model and adding NNCF optimization will only make it slower. So I would love to hear what you think about this.

- [x] rebase your work to include the #929 fix.
- [x] change the name of the notebook.

ashish-2005 commented:
docker_nbval still failing on gpt2 and whisper notebooks

adrianboguszewski commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
```diff
@@ -0,0 +1,666 @@
+{
```
<p>The images are too small. No one can read them, please enlarge :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

adrianboguszewski replied:
<pre class="ql-syntax" spellcheck="false">Line #23.        img = img.reshape([1,h,w,3])
</pre><p>The image is already in HWC. Just add a new dimension at the beginning</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

adrianboguszewski replied:
<pre class="ql-syntax" spellcheck="false">Line #5.    serialize(model_ir,'IR/efficientdet.xml')
</pre><p>Use path merging from Pathlib</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

adrianboguszewski replied:
<pre class="ql-syntax" spellcheck="false">Line #27.            output_keys = [model.output(4),  # CLASS KEY
</pre><p>They were already assigned. Why does it happen again?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

adrianboguszewski commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
```diff
@@ -0,0 +1,666 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #15.        if (img.shape[-1]) == 1:
</pre><p>We don't assume there are grayscale images. Let's use KISS rule.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

ashish-2005 replied:
I added this because there are some grayscale images in the COCO validation dataset, which were causing errors while validating the model. so should I skip grayscale images while validating?

adrianboguszewski replied:
I see. I think we can assume we will use BGR images only

adrianboguszewski reviewed (changes requested):
- There are many issues with punctuation in markdown and code. See the rules here: https://www.noslangues-ourlanguages.gc.ca/en/writing-tips-plus/punctuation-spacing
- More descriptions are required. We need something more than just bullets. You should tell the story, explaining what happens step by step. Every cell should be described
- If not committed now, remove the nncf part from readme and notebook
- Remove all HTML tags and replace them with markdown

ashish-2005 commented:
Hii,

I have updated the notebooks with asked changes

- [x] more description, still some bullet points in places where steps are explained
- [x] removed nncf optimization part
- [x] replaced all HTML with markdown
- [x] changes with respect to comments

ashish-2005 commented:
All CIs have passed except code_check for README.md's part I didn't change, after syncing my PR with the latest commits,
error:
```
FAILED .ci/test_notebooks.py::test_readme - AssertionError: notebooks/233-blip-visual-language-processing not found in notebooks README: README.md
assert 'notebooks/233-blip-visual-language-processing' in 'English | [简体中文](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/README_cn.md)\n\n<h1 align="center">...in-production/success-stories.html)\n\n\n---\n\n\\* Other names and brands may be claimed as the property of others.\n'
```

ashish-2005 commented:
Hii @adrianboguszewski 

I just want to confirm that, is there no README in the openvino_notebooks/notebooks directory now.

thanks

adrianboguszewski commented:
Yes, we removed it meanwhile

ashish-2005 commented:
CIs have failed because of some error while converting the TF model in SavedModel format to OpenVINO-IR. This error occurred after the latest updates and I have raised this issue in openvino repository. While this issue resolves I am trying to save the model graph (.pb) and use it to obtain IR, hoping it will work.

ashish-2005 commented:
Hii,

I have resolved the conversion issue, hence all CIs have passed except `build_docker` which failed on notebook 406 that I didn't alter. I ran 406 notebook both locally and on binder, it worked fine but somehow failed on CI.

therefore I think PR is ready for further review.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
```diff
@@ -0,0 +1,624 @@
+      "outputs": [],
+      "source": [
+        "# These dependencies will be added to the requirements.txt once approved.\n",
+        "!pip install openvino-dev[tensorflow2] tensorflow-hub pycocotools"
```
Thanks for contributing to the EfficientDet-optimization-notebook. I recommend moving the software requirements mentioned in the notebook to a separate file, such as requirements.txt , instead of having them in the notebook code itself. This will improve the clarity and maintainability of the notebook code.

ashish-2005 replied:
Correct me if I am wrong, but I think we are not allowed to have any other files except notebook and readme.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
```diff
@@ -0,0 +1,624 @@
+        "import numpy as np\n",
+        "\n",
+        "import tensorflow as tf\n",
+        "import tensorflow_hub as hub\n",
```
I am encountering an error while running the pre-requisites on my Windows machine. Could you please look into this issue and resolve it? 
Thank you.
<img width="617" alt="image" src="https://user-images.githubusercontent.com/109281183/231428570-94f618da-8353-4cf8-b777-29f1002891ae.png">

ashish-2005 replied:
I guess you might have duplicated version of TensorFlow in your environment.

AnishaUdayakumar replied:
It is recommended not to include such dependencies like "openvino-dev[tensorflow2]" within the notebook itself to avoid conflicts with the existing requirements.txt file. Instead, suggest adding it to the requirements.txt file. This will help ensure smooth integration and compatibility with the current notebook.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/README.md:
```diff
@@ -0,0 +1,20 @@
+Research Paper : [EfficientDet: Scalable and Efficient Object Detection](https://openaccess.thecvf.com/content_CVPR_2020/papers/Tan_EfficientDet_Scalable_and_Efficient_Object_Detection_CVPR_2020_paper.pdf)\
+Official GitHub : [EfficientDet](https://github.com/google/automl/tree/master/efficientdet)
+
+[EfficientDets](https://github.com/google/automl/tree/master/efficientdet) are a family of object detection models, which achieve state-of-the-art 55.1mAP on COCO test-dev, yet being 4x - 9x smaller and using 13x - 42x fewer FLOPs than previous detectors. EfficientDets also run 2x - 4x faster on GPU, and 5x - 11x faster on CPU than other detectors
```
It would be helpful to include a brief summary of what the notebook/tutorial demonstrates, following the guidelines for readme files. Providing an overview of the notebook content will improve its accessibility and help users determine if it meets their needs. Thank you for considering this suggestion.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/README.md:
```diff
@@ -0,0 +1,20 @@
+[EfficientDets](https://github.com/google/automl/tree/master/efficientdet) are a family of object detection models, which achieve state-of-the-art 55.1mAP on COCO test-dev, yet being 4x - 9x smaller and using 13x - 42x fewer FLOPs than previous detectors. EfficientDets also run 2x - 4x faster on GPU, and 5x - 11x faster on CPU than other detectors
+
+## Notebook Content
+
```
To improve the clarity and accessibility of the notebook content, it would be helpful to provide step-by-step instructions that explain the process in detail. Please refer to the guidelines for formatting and punctuation, to ensure the instructions are consistent and easy to follow.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
```diff
@@ -0,0 +1,624 @@
+        "id": "js4jU4zhchQ_"
+      },
+      "source": [
+        "# EfficientDet with OpenVINO™\n",
```
To improve the clarity and accuracy of the notebook title, it would be beneficial to revise it to something more specific and meaningful, such as 'Convert and Optimize EfficientDet with OpenVINO™' or something similar instead of 'EfficientDet with OpenVINO™'. This would help users understand the content and purpose of the notebook before opening it.

AnishaUdayakumar reviewed (changes requested):
Request to work on the suggestions.

AnishaUdayakumar commented on notebooks/240-efficientdet-object-detection/240-efficientdet-object-detection.ipynb:
To improve the consistency and clarity of the notebook content, it would be helpful to review and revise the titles of most of the Markdown cells. Please refer to the existing guidelines and notebooks for guidance on appropriate punctuation and terminology to ensure consistency throughout the notebook.

ashish-2005 commented:
Hi,

I have made the changes as suggested.

- [x] More descriptive Readme
- [x] Better markdown 
- [x] New title

Please re-review and suggest places for improvements, if any.
thanks

_(build_docker CI failed on notebooks I didn't alter)_

raymondlo84 commented:
# dependencies
!pip install openvino-dev[tensorflow2] 

I won't recommend installing big dependencies here. Maybe we shall comment them out.

ashish-2005 commented:
Hii @raymondlo84, thank you for your suggestion. Can you please expand it so I can start working on it? 
Like, which dependencies would you like me to avoid installing?

AnishaUdayakumar commented:
> Hii @AnishaUdayakumar I am not sure if you want me to create a new requirements.txt file in my tutorial repository or just add the dependency in the main existing requirements.txt (openvino_notebooks/requirements.txt). Please make it clear. Thanks.

I recommend adding the required dependencies to the main existing requirements.txt (openvino_notebooks/requirements.txt) instead of creating a new one. This will help ensure that users only need to install one file to get everything they need to run the tutorial.
Please let me know if you have any questions or if there's anything else I can help with.

AnishaUdayakumar commented:
> Hii @AnishaUdayakumar I am not sure if you want me to create a new requirements.txt file in my tutorial repository or just add the dependency in the main existing requirements.txt (openvino_notebooks/requirements.txt). Please make it clear. Thanks.

Hi @ashish-2005 , sorry for the confusion.
There seems to be some confusion regarding the dependencies used in the notebook. It is recommended to have the dependencies within the notebook itself.

In your notebook, I noticed that you have used general imports. To ensure that there are no conflicts while running your notebook, please provide specific versions of the dependencies that align with the Openvino_notebook/requirements.txt file.
Please note that the version you specify is aligned with the main requirements. For instance, instead of using the import statement "import tensorflow", please provide the specific version of TensorFlow used, such as "import tensorflow==X.X.0". Similarly, for "import tensorflow_hub", please provide the specific version, such as "import tensorflow_hub==0.XX.0".

Let me know if you have any questions

ashish-2005 commented:
Thanks, @AnishaUdayakumar for a detailed explanation. 
But I don't think it's possible to import a specific version as you mentioned, but it will get taken care of once the user installs the requirements.txt all the packages will get aligned. And for packages within the notebook, I will install the specific version to ensure they align with the main requirements.txt.

AnishaUdayakumar commented:
> @AnishaUdayakumar, is the PR ready?

No @adrianboguszewski , I have trouble running the notebook on my Windows machine. There is a conflict with the installation of some of the package versions like tensorflow, tensorflow_hub, protobuf, etc. I have been unable to resolve it and request some help.

ashish-2005 commented on notebooks/240-efficient-object-detection/240-efficient-object-detection.ipynb:
1) DONE.
2) All caps are used for constants like file path and URL as mentioned in guideline 9 [here](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/CONTRIBUTING.md#coding-guidelines).

ashish-2005 commented on notebooks/240-efficient-object-detection/240-efficient-object-detection.ipynb:
```diff
@@ -0,0 +1,615 @@
+      "outputs": [],
+      "source": [
+        "# dependencies\n",
+        "!pip install openvino-dev[tensorflow2]==2022.3.0 tensorflow-hub==0.13.0 pycocotools==2.0.6"
```
Installation of TF and protobuf is included. As of openvino-dev[tensorflow2] it is necessary for TF model conversion, as there is a bug that has been taken care of in [this PR](https://github.com/openvinotoolkit/openvino/pull/17311) but will be included in further release till then it's necessary to install openvino-dev[tensorflow2] within the notebook.

ashish-2005 commented on notebooks/240-efficient-object-detection/240-efficient-object-detection.ipynb:
```diff
@@ -0,0 +1,615 @@
+      },
+      "outputs": [],
+      "source": [
+        "import numpy as np\n",
+        "\n",
+        "import tensorflow as tf\n",
+        "import tensorflow_hub as hub\n",
+        "\n",
+        "import cv2\n",
+        "import matplotlib.pyplot as plt\n",
+        "\n",
+        "import pathlib\n",
+        "import json\n",
+        "import sys\n",
+        "from zipfile import ZipFile\n",
+        "\n",
+        "sys.path.append('../utils/')\n",
+        "from notebook_utils import load_image , download_file"
```
not sure what it means please expand...

ashish-2005 commented on notebooks/240-efficient-object-detection/240-efficient-object-detection.ipynb:
```diff
@@ -0,0 +1,615 @@
+            "text/plain": [
+              "<Figure size 600x600 with 1 Axes>"
+            ],
+            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeQAAAFwCAYAAABpWQisAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOz92bMlSX7fiX18i+Usd899r6y9u9FoNAgQAEEOySExHA1olGz4MAbT2Egmk8n0qH+Gb2PzNnrQ2LxohNEQBCiwmwAb6L27uqqyttz3u54lInzTg3vEOTczq7sg0WzwkG6Wy7333HMiwt1/y/f3/X1dxBgjr8fr8Xq8Hq/H6/F6/K865P/aF/B6vB6vx+vxerwer ...
```
I believe most of the context and purpose have been cleared in markdown cells.

ashish-2005 commented:
Hii @AnishaUdayakumar @adrianboguszewski,

All the issues have been addressed in the latest updates. As for the Windows machine hopefully, it will work now as it has passed all Windows CIs.

Thanks.

AnishaUdayakumar commented on notebooks/240-efficient-object-detection/240-efficient-object-detection.ipynb:
```diff
@@ -0,0 +1,612 @@
+            "text/plain": [
+              "<Figure size 600x600 with 1 Axes>"
+            ],
+            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeQAAAFwCAYAAABpWQisAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOz92bMlSX7fiX18i+Usd899r6y9u9FoNAgQAEEOySExHA1olGz4MAbT2Egmk8n0qH+Gb2PzNnrQ2LxohNEQBCiwmwAb6L27uqqyttz3u54lInzTg3vEOTczq7sg0WzwkG6Wy7333HMiwt1/y/f3/X1dxBgjr8fr8Xq8Hq/H6/F6/K865P/aF/B6vB6vx+vxerwer ...
```
Please mention that the first-time data download might take time and downloading speed depends on internet connection.

eaidova commented on notebooks/250-efficient-object-detection/250-efficient-object-detection.ipynb:
```diff
@@ -0,0 +1,622 @@
+{
```
<p>is it possible to provide charts credit</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

eaidova replied:
<p>information is slightly outdated. tensoflow models is also supported by direct loading now. In order to not overwrite this cool notebook. Can we remove this sentence and just say something like "For TF models in Saved Model format, we can convert it to OpenVINO IR format before"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

eaidova replied:
<pre class="ql-syntax" spellcheck="false">Line #4.    model_ir = mo.convert_model(saved_model_dir=str(MODEL_DIR_PATH))
</pre><p>serialize(model_ir,str(XML_PATH)) - please fix code style (not enough white space)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

eaidova replied:
<pre class="ql-syntax" spellcheck="false">Line #5.    model_ir = core.read_model(model=XML_PATH)
</pre><p>how do you decide when need to use named parameters and when not? </p><p>in read model line you use it, however only one parameter provided, but in compile_model not, however multiple parameters provided. Could you please align here and cell above?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

eaidova replied:
<p>&gt;To test model inference, create an inference pipeline with the help of the above-defined utility functions\  The pipeline consists of :- </p><p>I do not know, it is issue with my viewer, I see some strange symbols in this line (\ and -). Could you please remove?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/871/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

ashish-2005 commented:
Hi,

Apologies for no updates on this PR for such a long time. Do you still want me to finish this PR?

Thanks.

adrianboguszewski commented:
I'm closing this due to inactivity for a long time. Nevertheless, thank you for contributing and applying for Google Summer of Code. We encourage you to consider GSoC 2024 as well :)
